### PR TITLE
tests/raftstore: add a test case for cluster partition

### PR DESCRIPTION
**Pull request by tiancaiamao**

### CODE REVIEW

File: tests/raftstore/test_transport.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    } else {
+        (((count + 1) / 2..count + 1).collect(), (1..(count + 1) / 2).collect())
+    }
+}
+
+fn split_leader_with_minority(leader: u64, count: u64) -> (HashSet<u64>, HashSet<u64>) {
+    let mut vec: Vec<u64> = (1..count + 1).collect();
+    vec.swap(0, (leader - 1) as usize);
+    let ucount = count as usize;
+    let s1: HashSet<_> = vec[0..(ucount - 1) / 2].into_iter().cloned().collect();
```
Comment (siddontang) [inline tests/raftstore/test_transport.rs]: can remove` HashSet<_>`?

File: tests/raftstore/test_transport.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    let s2: HashSet<_> = vec[(ucount - 1) / 2..ucount].into_iter().cloned().collect();
+    (s1, s2)
+}
+
+fn test_partition_write<T: Simulator>(cluster: &mut Cluster<T>, count: u64) {
+    cluster.bootstrap_region().expect("");
+    cluster.start();
+
+    let (key, value) = (b"k1", b"v1");
+    cluster.must_delete(key);
```
Comment (siddontang) [inline tests/raftstore/test_transport.rs]: why delete here? these is no data here.

File: tests/raftstore/test_transport.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    let region_id = cluster.get_region_id(key);
+    let peer = cluster.leader_of_region(region_id).unwrap();
+    let leader = peer.get_store_id();
+
+    // leader partition with majority doesn't affect write/read
+    let (s1, s2) = split_leader_with_majority(leader, count);
```
Comment (siddontang) [inline tests/raftstore/test_transport.rs]: check leader changed after majority split.

File: tests/raftstore/test_transport.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    cluster.partition(Arc::new(s1), Arc::new(s2));
+    cluster.must_put(key, value);
+    assert_eq!(cluster.get(key), Some(value.to_vec()));
+    cluster.reset_transport_hooks();
+
+    // if leader partition with minority, new leader should be elected
+    let (s1, s2) = split_leader_with_minority(leader, count);
```
Comment (siddontang) [inline tests/raftstore/test_transport.rs]: check leader still the same after minority split.

File: tests/raftstore/test_transport.rs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    cluster.bootstrap_region().expect("");
+    cluster.start();
+
+    let (key, value) = (b"k1", b"v1");
+    let region_id = cluster.get_region_id(key);
+    let leader = cluster.leader_of_region(region_id).unwrap();
+
+    // leader in majority, partition doesn't affect write/read
+    let (s1, s2) = leader_in_majority_split(leader, count);
+    cluster.partition(Arc::new(s1), Arc::new(s2));
```
Comment (ngaut) [inline tests/raftstore/test_transport.rs]: Can we simply write the Api like this: cluster.partition(vec![1,2,leader], [3,4]) ?

Comment (tiancaiamao) [inline tests/raftstore/test_transport.rs]: if we know 5 is leader, of course this is no problem.
but I know what you mean is something else...that's much difficult and the convenience it brings just not deserve a try.

Comment (ngaut) [inline tests/raftstore/test_transport.rs]: Yep, that's why we need leadership transfer :)

**Comment by siddontang** [comment]
LGTM

**Comment by ngaut** [comment]
We should wait for https://github.com/pingcap/tikv/pull/534 and clean up the testing API.

**Comment by siddontang** [comment]
Got it

**Comment by tiancaiamao** [comment]
updated... PTAL @ngaut

**Comment by ngaut** [comment]
LGTM

**Pull request merged**